pow/src/compute/mod.rs

```diff
@@ -89,34 +92,82 @@ fn need_new_vm<M: randomx::WithCacheMode>(
 	need_new_vm
 }
 
+fn do_new_vm<M: randomx::WithCacheMode>(
+	key_hash: &H256,
+	machine: &RefCell<Option<(H256, randomx::VM<M>)>>,
+	shared_caches: &Mutex<LruCache<H256, Arc<randomx::Cache<M>>>>,
+	f_has_large_pages: fn(&Config) -> bool,
+) -> Result<(), ()> {
+	let mut shared_caches = shared_caches.lock();
+
+	if !f_has_large_pages(global_config()) {
+		if let Some(cache) = shared_caches.get_mut(key_hash) {
+			machine.replace(Some((*key_hash, randomx::VM::new(cache.clone(), global_config()))));
+
+			Ok(())
+		} else {
+			info!(
+				target: "kulupu-randomx",
+				"At block boundary, generating new RandomX {} cache with key hash {} ...",
+				M::description(),
+				key_hash,
+			);
+
+			let cache = match randomx::Cache::new(&key_hash[..], global_config()) {
+				Ok(cache) => Arc::new(cache),
+				Err(_) => panic!("Cache allocation failed"),
```

Comment:
Please remove those panic and return appropriate errors.

In particular, we should not panic because of randomx failures. Those errors should propagate back to the sender so that the client can gracefully shutdown.